### ISSUE TITLE
`ColumnSelector ` should also be able to access globals

### ISSUE THREAD

**Issue by yych42** [yych42/letsgo#3]
See the screenshot below. It's a good idea to define underlying data types like `dataset`. We should make sure nodes that accepts a certain dataset can always access it no matter where it originates from. This means the `ColumnSelector` node should be able to access higher-level globals when used in a tree structure, and adapt to data provided by source nodes, such as selected column, when there's none currently active.

![CleanShot 2024-05-16 at 18 34 53@2x](https://github.com/yych42/letsgo/assets/43839115/6167277f-a649-44c1-8b2d-0398b367816e)

**Comment by yych42** [comment]
I fixed the above issue with #4, but see comments for some other related problems.

Also identified a new issue:

![CleanShot 2024-05-16 at 18 57 00@2x](https://github.com/yych42/letsgo/assets/43839115/14c0de21-e183-49ee-96d0-72e6dd3f3cbb)

Generally, if a node provides access to a kind of data type at some point in the node tree, it should always allow for access in the subsequent nodes down the tree. This makes me think I am a bit confused about how globals are supposed to work, since sometimes a node operation (e.g., `RangeFilter`) can introduce new datasets that subsequent nodes should handle it just as well. 

Perhaps when a node creates a new dataset, it should also create a new global? But I feel like this could complicate down-the-tree handling, when you need to locate the right global to access from a whole dictionary of globals. A solution for this is to also make these nodes replace the `globals` data (perhaps making sure the node only holds one `global` instead of multiple), so all the down-the-tree nodes only have to worry about one global. By this design, "globals" probably shouldn't be called so. 

Maybe something like "checkpoint" describes the role better. Some nodes are descriptive: they need a provided dataset to display some information (perhaps also passing it down if needed; e.g., ColumnNames) but they don't need to change the provided dataset, so they just borrow the inflowing dataset, do what they need to do, and pass down a checkpoint for subsequent nodes (along with whatever thing they added). Some nodes may have altered the dataset, so they create and hold a copy themselves, and update the checkpoint so subsequent nodes access this latest checkpoint.

Another idea I was thinking about is for certain nodes to only aggregate "operational steps" along the path. The `RangeFilter` could be one of these nodes; all they do is to annotate the steps, without actually doing any work. This allows for multiple filters to be chained together, without actually creating a bunch of slightly different datasets. Consider the following setup:

![CleanShot 2024-05-16 at 20 05 51@2x](https://github.com/yych42/letsgo/assets/43839115/102f4e77-cc80-406c-bf39-f18fe66d44da)

(Note that the last `ColumnSelector` node is currently not functional due to the same issue.)

In this flow, no node (except for `Inspector`) actually needed to know anything about the updated dataset(s), so why bother computing and storing those? The whole thing can be represented as such:

```json
{
  "checkpoint": {
    "at": "some_node_id",
    "type": "dataset",
    "queued_steps": [
      {
        "by": "some_node_id",
        "target_column": "var2",
        "operation": "filter_by_range",
        "parameters": {
          "min": 0,
          "max": 10,
          "set_as_missing": true
        }
      },
      {
        "by": "some_node_id",
        "target_column": "var1",
        "operation": "filter_by_range",
        "parameters": {
          "min": 1,
          "max": 6,
          "set_as_missing": true
        }
      }
    ]
  }
}
```

Every node knows which column they need to read, so they can check and see if there are any pending operations. If not, they just read from the last checkpoint and do whatever they need to do (e.g., calculate some descriptives in the case of `ColumnSelector`). But, let's say the last `ColumnSelector` looks across the `queued_steps` and realized it was selecting `var2`, something that was supposed to be updated already. Then, this node executes all the queued steps, start holding a new dataset, and update the checkpoint flag to point to itself, and remove _all_ existing steps since they are now realized. The same works for a `ExportCSV` node. I guess the idea is all the nodes has access to some kind of an executor, which could deal with all the dataset-altering operations that could show up in these operational steps, and if it is relevant, it ask the executor to execute the changes and get a new copy of the dataset that it needs.

Or maybe I misunderstood the whole thing.

At least we should try to define the data that a validly connected node should always provide.

@runkaiz

**Comment by yych42** [comment]
Since I'm back to solo dev, I'll merge the temporary fix for now, and explore different solutions as I go. Maybe it's just too early to optimize for this.
